pkg/model/mselection/heaviest.go

```diff
@@ -340,10 +297,46 @@ func (s *HeaviestSelector) tipItems() []*item {
 	return result
 }
 
-// randomTip selects a random tip from the provided slice of tips.
-func randomTip(tips hornet.Hashes) hornet.Hash {
+// GetStats returns the amount of known tips and approvees of s.
+func (s *HeaviestSelector) GetStats() (tipCount int, approveeCount int) {
+	return s.tips.Len(), len(s.approvers)
+}
+
+// randomTipFromTips selects a random tip from the provided slice of tips.
+func randomTipFromTips(tips []*selectedTip) (*selectedTip, error) {
+	if len(tips) == 0 {
+		return nil, ErrNoTipsAvailable
+	}
+
+	return tips[utils.RandomInsecure(0, len(tips)-1)], nil
+}
+
+// randomTipFromItems selects a random tip item from the provided slice of items.
+func randomTipFromItems(tips []*item) (*selectedTip, error) {
 	if len(tips) == 0 {
-		panic("empty tips")
+		return nil, ErrNoTipsAvailable
+	}
+
+	index := utils.RandomInsecure(0, len(tips)-1)
+	return &selectedTip{item: tips[index], index: index}, nil
+}
+
+// applyTip set all bits of all referenced transactions of the tip in all existing tips to zero.
+func applyTip(tips []*item, tip *selectedTip) []*item {
```

Comment:
I feel like the list of items should be an own type with operations on it, instead of helper functions where you pass the list